// @mui/x-data-grid/node/hooks/features/rows/useGridRowsMeta.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useGridRowsMeta = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var React = _interopRequireWildcard(require("react"));

var _useCurrentPageRows = require("../../utils/useCurrentPageRows");

var _useGridApiMethod = require("../../utils/useGridApiMethod");

var _useGridSelector = require("../../utils/useGridSelector");

var _densitySelector = require("../density/densitySelector");

var _gridFilterSelector = require("../filter/gridFilterSelector");

var _gridPaginationSelector = require("../pagination/gridPaginationSelector");

var _gridSortingSelector = require("../sorting/gridSortingSelector");

var _useGridStateInit = require("../../utils/useGridStateInit");

var _gridEvents = require("../../../models/events/gridEvents");

var _useGridApiEventHandler = require("../../utils/useGridApiEventHandler");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

/**
 * @requires useGridPageSize (method)
 * @requires useGridPage (method)
 */
const useGridRowsMeta = (apiRef, props) => {
  const {
    getRowHeight,
    pagination,
    paginationMode
  } = props;
  const rowsHeightLookup = React.useRef({});
  const rowHeight = (0, _useGridSelector.useGridSelector)(apiRef, _densitySelector.gridDensityRowHeightSelector);
  const filterState = (0, _useGridSelector.useGridSelector)(apiRef, _gridFilterSelector.gridFilterStateSelector);
  const paginationState = (0, _useGridSelector.useGridSelector)(apiRef, _gridPaginationSelector.gridPaginationSelector);
  const sortingState = (0, _useGridSelector.useGridSelector)(apiRef, _gridSortingSelector.gridSortingStateSelector);
  (0, _useGridStateInit.useGridStateInit)(apiRef, state => (0, _extends2.default)({}, state, {
    rowsMeta: {
      currentPageTotalHeight: 0,
      positions: []
    }
  }));
  const hydrateRowsMeta = React.useCallback(() => {
    const {
      rows
    } = (0, _useCurrentPageRows.getCurrentPageRows)(apiRef, {
      pagination,
      paginationMode
    });
    apiRef.current.setState(state => {
      const positions = [];
      const densityFactor = (0, _densitySelector.gridDensityFactorSelector)(state, apiRef.current.instanceId);
      const currentRowHeight = (0, _densitySelector.gridDensityRowHeightSelector)(state, apiRef.current.instanceId);
      const currentPageTotalHeight = rows.reduce((acc, row) => {
        positions.push(acc);
        let baseRowHeight;
        const isResized = rowsHeightLookup.current[row.id] && rowsHeightLookup.current[row.id].isResized || false;

        if (isResized) {
          // do not recalculate resized row height and use the value from the lookup
          baseRowHeight = rowsHeightLookup.current[row.id].value;
        } else {
          baseRowHeight = currentRowHeight;

          if (getRowHeight) {
            var _getRowHeight;

            // Default back to base rowHeight if getRowHeight returns null or undefined.
            baseRowHeight = (_getRowHeight = getRowHeight((0, _extends2.default)({}, row, {
              densityFactor
            }))) != null ? _getRowHeight : currentRowHeight;
          }
        }

        const heights = apiRef.current.unstable_applyPreProcessors('rowHeight', {
          base: baseRowHeight
        }, // We use an object to make simple to check if a size was already added or not
        row);
        const finalRowHeight = Object.values(heights).reduce((acc2, value) => acc2 + value, 0);
        rowsHeightLookup.current[row.id] = {
          value: baseRowHeight,
          isResized
        };
        return acc + finalRowHeight;
      }, 0);
      return (0, _extends2.default)({}, state, {
        rowsMeta: {
          currentPageTotalHeight,
          positions
        }
      });
    });
    apiRef.current.forceUpdate();
  }, [apiRef, pagination, paginationMode, getRowHeight]);

  const getTargetRowHeight = rowId => {
    var _rowsHeightLookup$cur;

    return ((_rowsHeightLookup$cur = rowsHeightLookup.current[rowId]) == null ? void 0 : _rowsHeightLookup$cur.value) || rowHeight;
  };

  const setRowHeight = React.useCallback((id, height) => {
    rowsHeightLookup.current[id] = {
      value: height,
      isResized: true
    };
    hydrateRowsMeta();
  }, [hydrateRowsMeta]); // The effect is used to build the rows meta data - currentPageTotalHeight and positions.
  // Because of variable row height this is needed for the virtualization

  React.useEffect(() => {
    hydrateRowsMeta();
  }, [rowHeight, filterState, paginationState, sortingState, hydrateRowsMeta]);
  const handlePreProcessorRegister = React.useCallback(name => {
    if (name !== 'rowHeight') {
      return;
    }

    hydrateRowsMeta();
  }, [hydrateRowsMeta]);
  (0, _useGridApiEventHandler.useGridApiEventHandler)(apiRef, _gridEvents.GridEvents.preProcessorRegister, handlePreProcessorRegister);
  const rowsMetaApi = {
    unstable_getRowHeight: getTargetRowHeight,
    unstable_setRowHeight: setRowHeight
  };
  (0, _useGridApiMethod.useGridApiMethod)(apiRef, rowsMetaApi, 'GridRowsMetaApi');
};

exports.useGridRowsMeta = useGridRowsMeta;
